tools/modify_master.py

```diff
@@ -13,98 +28,171 @@
+        self.master_envvars = {}
+        self.commented_lines = []
+
+    def __enter__(self):
+        #Get the envvars at the start
+        self._fetch_current_envvars()
+        return self
+
+    def __exit__(self, *stuff):
+        _ = stuff
+        self._write_envvars()
+        self.restart_master()
+        logger.info("Modification complete.")
+
+    def set_master_envvar(self, envvar, value):
+        """Add an environment variable from the mesos master context
+
+        :param envvar: The envvar name to set
+        :type envvar: str
+        :param value: the value to use for envvar
+        :type value: str
+        """
+        logger.info("Setting %s=%s", envvar, value)
+        envvars = {envvar: value}
+        self.set_master_envvars(envvars)
+
+    def set_master_envvars(self, envvars):
+        """Set multiple environment variables in the mesos master context
+
+        :param envvars: A dictionary of envvar names and values
+        :type envvars: dict
+        """
+        logger.info("Setting envvars: %s", envvars)
+        def set_vars(previous_envvars, updates=envvars):
+            new_envvars = previous_envvars.copy()
+            new_envvars.update(updates)
+            return new_envvars
+        self._modify_envvars(set_vars)
+
+
+    def remove_master_envvar(self, envvar):
+        """Remove an an environment variable from the mesos master context
+
+        :param envvar: The envvar name to remove
+        :type envvar: str
+        """
+        logger.info("Unsetting %s", envvar)
+        def remove_var(previous_envvars, remove_var=envvar):
+            new_envvars = previous_envvars.copy()
+            del new_envvars[remove_var]
+            return new_envvars
+        self._modify_envvars(remove_var)
+
+
+    def _fetch_current_envvars(self):
+        """Acquire the curent set of envvars from the mesos master and store
+        them in self.master_envvars
+        """
+        # Get the current set of master envvars
+        success, out = shakedown.run_command_on_master('cat /opt/mesosphere/etc/mesos-master')
+        if success is not True:
+            logger.info('Unable to get current envvars from master: {}'.format(out))
+            raise RuntimeError("Unable to get current envvars")
+
+        logger.info("Current envvars:\n{}".format(out))
+        envvars, commented_lines = self._decode_master_config(out)
+        self.master_envvars = envvars
+        self.commented_lines = commented_lines
+
+    def _modify_envvars(self, modifier_f):
+        """Performs a transformation on the mesos master environment
+        variables.  Accepts a funciton which accepts the mesos environment
+        before the transformation, and returns the new environment.  The
+        function should accept and return dicts of str:str
+
+        :param modifier_f: The function to apply the transformation to the envvar state
+        :type modifier_f: func
+        """
```

Comment:
We could modify the dict in place, but it felt gross to me to have the functor dustructively modifying an object it doesn't really know the lifetime of.
I could move the protection against this to _modify_envvars(), or i could remove, but this felt more right to me than a destructive change that could break on refactor in weird ways.